scipy/stats/_unuran/unuran_wrapper.pyx.templ

```diff
@@ -1370,6 +1401,526 @@ cdef class NumericalInversePolynomial(Method):
+                  tol=None,
+                  construction_points=None,
+                  max_intervals=100000,
+                  guide_factor=1,
+                  random_state=None):
+        domain, order, u_resolution = self._validate_args(dist, domain, order, max_intervals,
+                                                          u_resolution, tol, construction_points)
+
+        # save all the arguments for pickling support
+        self._kwargs = {
+            'dist': dist,
+            'domain': domain,
+            'order': order,
+            'u_resolution': u_resolution,
+            'tol': tol,
+            'max_intervals': max_intervals,
+            'guide_factor': guide_factor,
+            'construction_points': construction_points,
+            'random_state': random_state
+        }
+
+        cdef:
+            unur_distr *distr
+            unur_par *par
+
+        self.callbacks = _unpack_dist(dist, "cont", meths=["cdf"], optional_meths=["pdf", "dpdf"])
+        def _callback_wrapper(x, name):
+            return self.callbacks[name](x)
+        self._callback_wrapper = _callback_wrapper
+        self._messages = MessageStream()
+        _lock.acquire()
+        try:
+            unur_set_stream(self._messages.handle)
+
+            self.distr = unur_distr_cont_new()
+            if self.distr == NULL:
+                raise UNURANError(self._messages.get())
+            _pack_distr(self.distr, self.callbacks)
+
+            if domain is not None:
+                self._check_errorcode(unur_distr_cont_set_domain(self.distr, domain[0],
+                                                                 domain[1]))
+
+            self.par = unur_hinv_new(self.distr)
+            if self.par == NULL:
+                raise UNURANError(self._messages.get())
+
+            self._check_errorcode(unur_hinv_set_order(self.par, order))
+            self._check_errorcode(unur_hinv_set_u_resolution(self.par, u_resolution))
+            self._check_errorcode(unur_hinv_set_max_intervals(self.par, max_intervals))
+            self._check_errorcode(unur_hinv_set_cpoints(self.par, &self.construction_points_array[0],
+                                                        len(self.construction_points_array)))
+            self._check_errorcode(unur_hinv_set_guidefactor(self.par, guide_factor))
+
+            self._set_rng(random_state)
+        finally:
+            _lock.release()
+
+    def _validate_args(self, dist, domain, order, max_intervals, u_resolution, tol,
+                       construction_points):
+        domain = _validate_domain(domain, dist)
+        # UNU.RAN raises warning and sets a default value. Prefer an error instead.
+        if order not in {1, 3, 5}:
+            raise ValueError("`order` must be either 1, 3, or 5.")
+        if int(max_intervals) != max_intervals or max_intervals <= 1:
+            raise ValueError("`max_intervals' must be an integer greater than 1.")
+        if tol is not None:
```

Comment:
If both tol and u_resolution are provided, should a warning be raised that u_resolution will be considered and tol should not be used?

Reply:
OK, I think you removed `tol` from the documented API and ensured that the value can still be provided. I think we might need to add a deprecation warning but not sure about the process

Reply:
> OK, I think you removed `tol` from the documented API and ensured that the value can still be provided. I think we might need to add a deprecation warning but not sure about the process

I haven't raised a deprecation warning because this is a doc-deprecation. But we can probably warn the users in the docs by saying that the constructor also accepts `tol` which is equivalent to `u_resolution` but `u_resolution` should be preferred over `tol` and both the arguments should not be passed at the same time.

Reply:
> but it seems to me like the default is to fall back on `tol`.

Yes, I noticed this a couple of days ago and had a chat with @chrisb83. The problem is that `u_resolution` has a non-none default value and there is no way to know when the user passed both `tol` and `u_resolution` i.e. there is no way to know the difference between this:

```py
>>> rng = NumericalInverseHermite(dist, tol=1e-12)
>>> rng = NumericalInverseHermite(dist, u_resolution=1e-10, tol=1e-12)
>>> # in the latter case, did the user pass both u_resolution and tol?
>>> # If so, how to tell that as u_resolution already has a default value of 1e-10
```

One way to deal with this is to keep `u_resolution=None` in the constructor and assign it the value `1e-10` if the user hasn't passed anything. This way, if both `u_resolution` and `tol` are not none, we could raise a warning and prefer `u_resolution` over `tol`. Does this make sense?

Reply:
I see the issue that you're dealing with, I'm not aware of a clean way to tell if the keyword default was supplied or not (especially since this is an immutable type).  Setting the default to `None` for `u-resolution` feels like an awkward solution for a new keyword. It could also cause confusion -- similar to how the behavior when `axis=None` in some functions is drastically different than when it is numerically set. 

The folks who will be using the `u_res` parameter will be future implementations, and they will know (or at least be able to find out if you raise a warning) that `tol` is being deprecated if they try to use it. Maintaining backwards compatibility should be prioritized, so I think there is only one real way forward, which is to use `tol` if it has been set. 

the two current usages for `tol` seem to be: 
1. Any previous users that did not pass in a value for `tol` will see the functionality maintained. The `u res` default value matches the previous `tol` value so this already works well.
2. users who did pass in a `tol ` (default `1e-12`  or other value) need to still have the same results regardless of what `u-res` is. 

Trying to force `u-res` to be preferred over `tol` when both are provided seems overly complicated for a scenario where the user is not reading the documentation. I think there would be a smoother path to being merged into main if backwards compatibility is maintained as closely as possible. 

In the end, it seems to me that if `tol` is supplied, you should indicate in a warning that while its input will be honored, it has been replaced with `u_resolution` in scipy 1.x, and it will may be fully deprecated in future releases. 

Others may have different opinions, though!

Reply:
> In the end, it seems to me that if tol is supplied, you should indicate in a warning that while its input will be honored, it has been replaced with u_resolution in scipy 1.x, and it will may be fully deprecated in future releases.

I agree, this seems to be the best option we have.

Reply:
Thanks for the detailed reply, @swallan!

> The folks who will be using the `u_res` parameter will be future implementations, and they will know (or at least be able to find out if you raise a warning) that `tol` is being deprecated if they try to use it. Maintaining backwards compatibility should be prioritized, so I think there is only one real way forward, which is to use `tol` if it has been set.

Nice catch! I didn't see it this way.

> Trying to force `u-res` to be preferred over `tol` when both are provided seems overly complicated for a scenario where the user is not reading the documentation. I think there would be a smoother path to being merged into main if backwards compatibility is maintained as closely as possible.

I agree. But I wanted to change the default value of `u_resolution` to `1e-10` instead of `1e-12` because UNU.RAN has that as default. This should not break backward compatibility because if the setup succeeded for `u_resolution=1e-12`, it should also succeed for `u_resolution=1e-10`. Moreover, `1e-10` is sufficiently close to machine precision and the setup is faster. This change doesn't affect the user so it seems reasonable to me. Do you agree?

> In the end, it seems to me that if `tol` is supplied, you should indicate in a warning that while its input will be honored, it has been replaced with `u_resolution` in scipy 1.x, and it will may be fully deprecated in future releases

This seems reasonable to me. I will raise a deprecation warning if `tol` is supplied.

Reply:
> This change doesn't affect the user 

I made a quick comparison to examine the average difference for these two scenarios:
- previous implementation => new implementation both at 1e-12
- previous implementation @ 1e-12 => new implementation at 1e-10

and then found the bootstrapped confidence intervals for mean absolute error for both:
```
stats.bootstrap([np.abs(diff_12_to_10)], np.mean, confidence_level=.99)
Out[14]: BootstrapResult(confidence_interval=ConfidenceInterval(low=1.448447252711894e-10, high=1.1200871877506022e-09), standard_error=1.630491764207449e-10)

stats.bootstrap([np.abs(diff_both_12)], np.mean, confidence_level=.99)
Out[15]: BootstrapResult(confidence_interval=ConfidenceInterval(low=1.0504267543195413e-12, high=5.4826531970922883e-11), standard_error=8.617733140441905e-12)
```

So the difference between going from previous code @ 1e-12 to new code at 1e-10 has a slightly larger absolute error between the samples. Small, but possibly something to think about. I'm not completely familiar if the results are supposed to match up exactly, so if they are not, disregard.

Reply:
Thanks for the analysis!

> So the difference between going from previous code @ 1e-12 to new code at 1e-10 has a slightly larger absolute error between the samples than going from previous to new with both at 1e-12. Small, but possibly something to think about.

I think this is expected because we are moving to a higher error bound. I don't think this change should affect the user very much as it is too small. What do you think @chrisb83?

Reply:
Even if we keep the error at 1e-12, there will be a little bit of variation in the random variates due to the change from the Python implementation to UNU.RAN (Samuel's analysis shows this). Changing the u-resolution from 1e-12 to 1e-10 will add a bit of noise again (to isolate the effect, it would be best to compare the same generator with the different values of `u_error`)

From a practical point of view, this should be irrelevant: if you use the simulated rvs in a simulation, differences of order 1e-9 in the underlying rvs should not matter at all.

Reply:
Thanks @chrisb83 and @swallan! I have fixed the direction warnings and added one for `tol`. Let me know if there are any other points left to address.